src/chainstate/coordinator/mod.rs

```diff
@@ -0,0 +1,380 @@
+                        if is_pox_anchor == true {
+                            // todo: can call this method when the BlockSnapshot struct is augmented with block_id field
+                            // self.process_new_pox_anchor(block.block_id)?;
+                        }
+                        Ok(())
+                    }
+                    Err(err) => Err(err)
+                }?;
+                self.canonical_chain_tip = Some(block);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn process_new_pox_anchor(&mut self, block_id: &BlockIdentifier) -> Result<(), ()> {
+        // Ensure that the chain of anchored blocks (up to block_id) has been processed  
+        let ordered_missing_anchored_blocks = self.pox_db.get_ordered_missing_anchors(block_id);
+        for block_id in ordered_missing_anchored_blocks.iter() {
+            match self.chain_state_db.is_block_processed(&block_id) {
+                Ok(is_processed) => {
+                    if is_processed {
+                        PoxDB::process_anchor(&block_id, &self.chain_state_db)?;
+                        let canonical_bhh = self.pox_db.get_reward_set_start_for(&block_id);
+                        // Retrieve the corresponding block
+                        let canonical_block = self.burnchain_blocks_db.get_burnchain_block(canonical_bhh)?;
+                        self.canonical_burnchain_chain_tip = Some(canonical_block); 
+                        self.canonical_pox_id = Some(self.pox_db.get_canonical_pox_id(&block_id.burnchain_header_hash));
+                        self.canonical_chain_tip = None;
+                        // Start processing from the beginning of the new PoX reward set
+                        self.handle_new_burnchain_block()?;    
+                    }
+                    Ok(())
+                },
+                Err(e) => {
+                    error!("ChainsCoordinator: unable to retrieve processed block {:?}", block_id);
+                    Err(e)
+                }
+            }?;
+        }
+        self.discover_new_pox_anchor(block_id)
+    }
+
+    #[allow(unused_variables)]
+    fn process_block(&self, block: &Block) -> Result<bool, ()> {
+        unimplemented!()
+    }
+
+    fn discover_new_pox_anchor(&mut self, block_id: &BlockIdentifier) -> Result<(), ()> {
+        PoxDB::process_anchor(block_id, &self.chain_state_db)
+    }
+}
+
+pub enum ChainsEvent {
+    BlockDiscovered(Option<Sender<ChainsEventCallback>>),
+    BurnchainBlockDiscovered(Option<Sender<ChainsEventCallback>>),
+}
+
+pub enum ChainsEventCallback {
+    BlockProcessed,
+    BurnchainBlockProcessed,
+}
+
+pub struct ChainsEventsRouter {
+    events_tx: Option<mpsc::Sender<ChainsEvent>>,
+}
+
+impl ChainsEventsRouter {
+
+    pub fn new() -> ChainsEventsRouter {
+        ChainsEventsRouter {
+            events_tx: None,
+        }
+    }
+
+    pub fn spawn_chains_coordinator(&mut self) -> mpsc::Sender<ChainsEvent> {
```

Comment:
This method looks like it's spawning the "main loop" in a separate thread.  Can we instead make it so the "main loop" _is_ the main thread, and all other processes (networking, block downloading, processing, etc.) get spawned as threads?

Reply:
Yeah, I don't really know at this point how and who will be instantiating the `ChainsCoordinator`, so the threading strategy is implemented in the struct `ChainsEventsRouter`, not the `ChainsCoordinator`. Meaning when we'll be plugging the `ChainsCoordinator` with the node, we can either use the struct `ChainsEventsRouter`, or take the code from this struct and just get rid it (if it's not being used in our tests).

Reply:
Let's assume for now that the caller will be responsible for instantiating the thread, and have this method just implement the thread body.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1704/commits/fbc867ab05a98fe304f4ecd895ba05e172368716